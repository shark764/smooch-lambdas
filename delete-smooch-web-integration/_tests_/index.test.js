const { lambda: { api: { validateTenantPermissions } } } = require('alonzo');

jest.mock('aws-sdk');
jest.mock('alonzo');

validateTenantPermissions.mockReturnValue(true);

const event = {
  params: {
    'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
    id: '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
  identity: {
    'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
  },
};

const mockGetSecretValue = jest.fn(() => {})
  .mockImplementation(() => ({
    promise: () => ({
      SecretString: JSON.stringify({
        '5e31c81640a22c000f5d7f28-id': 'id',
        '5e31c81640a22c000f5d7f28-secret': 'secret',
      }),
    }),
  }));

const mockGet = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({
      Item: { 'app-id': '5e31c81640a22c000f5d7f28' },
    }),
  }));

const mockDeleteSmoochCore = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

const mockDelete = jest.fn()
  .mockImplementation(() => ({
    promise: () => ({}),
  }));

const mockSmoochCore = jest.fn(() => ({
  integrations: {
    delete: mockDeleteSmoochCore,
  },
}));

jest.mock('aws-sdk', () => ({
  config: {
    update: jest.fn(),
  },
  DynamoDB: {
    DocumentClient: jest.fn().mockImplementation(() => ({
      get: mockGet,
      delete: mockDelete,
    })),
  },
  SecretsManager: jest.fn().mockImplementation(() => ({ getSecretValue: mockGetSecretValue })),
}));

jest.mock('smooch-core', () => mockSmoochCore);

const { handler } = require('../index');

describe('delete-smooch-web-integration', () => {
  describe('Everthing is successful', () => {
    it('sends back status 200 when the web integration is deleted successfully', async () => {
      const result = await handler(event);
      expect(result).toMatchSnapshot();
    });
    describe('Walkthrough', () => {
      beforeAll(async () => {
        await handler(event);
      });
      it('passes in the correct arguments to secretClient.getSecretValue()', async () => {
        expect(mockGetSecretValue.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to validateTenantPermissions', async () => {
        expect(validateTenantPermissions.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to docClient.get()', async () => {
        expect(mockGet.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to SmoochCore', async () => {
        expect(mockSmoochCore.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to smooch.integrations.delete()', async () => {
        expect(mockDeleteSmoochCore.mock.calls).toMatchSnapshot();
      });

      it('passes in the correct arguments to docClient.delete()', async () => {
        expect(mockDelete.mock.calls).toMatchSnapshot();
      });
    });
  });
  it('sends back status 400 when there are invalid parameters', async () => {
    const mockEvent = {
      params: {
        'tenant-id': '66d83870-30df-4a3b-8801-59edff162034',
      },
      identity: {
        'user-id': '667802d8-2260-436c-958a-2ee0f71f73f0',
      },
    };
    const result = await handler(mockEvent);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error retrieving digital channels credentials', async () => {
    mockGetSecretValue.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 403 when there are not enough permissions', async () => {
    validateTenantPermissions.mockReturnValueOnce(false);
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when the app does not exist for the tenant', async () => {
    mockGet.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error fetching an app in DynamoDB', async () => {
    mockGet.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error validating digital channels credentials', async () => {
    mockGetSecretValue.mockImplementationOnce(() => ({
      promise: () => ({}),
    }));
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error deleting web integration', async () => {
    mockDeleteSmoochCore.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });

  it('sends back status 500 when there is a error deleting records in DynamoDB', async () => {
    mockDelete.mockRejectedValueOnce(new Error());
    const result = await handler(event);
    expect(result).toMatchSnapshot();
  });
});
