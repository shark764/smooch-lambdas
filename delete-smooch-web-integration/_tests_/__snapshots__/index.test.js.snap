// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to SmoochCore 1`] = `
Array [
  Array [
    Object {
      "keyId": "id",
      "scope": "app",
      "secret": "secret",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
  Array [
    Object {
      "keyId": "id",
      "scope": "app",
      "secret": "secret",
      "serviceUrl": "mock-smooch-api-url",
    },
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to docClient.delete() 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "id": "667802d8-2260-436c-958a-2ee0f71f73f0",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
  Array [
    Object {
      "Key": Object {
        "id": "667802d8-2260-436c-958a-2ee0f71f73f0",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to docClient.get() 1`] = `
Array [
  Array [
    Object {
      "Key": Object {
        "id": "667802d8-2260-436c-958a-2ee0f71f73f0",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
  Array [
    Object {
      "Key": Object {
        "id": "667802d8-2260-436c-958a-2ee0f71f73f0",
        "tenant-id": "66d83870-30df-4a3b-8801-59edff162034",
      },
      "TableName": "us-east-1-dev-smooch",
    },
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to secretClient.getSecretValue() 1`] = `
Array [
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
  Array [
    Object {
      "SecretId": "us-east-1-dev-smooch-app",
    },
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to smooch.integrations.delete() 1`] = `
Array [
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "integrationId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
  ],
  Array [
    Object {
      "appId": "5e31c81640a22c000f5d7f28",
      "integrationId": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful Walkthrough passes in the correct arguments to validateTenantPermissions 1`] = `
Array [
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "WEB_INTEGRATIONS_APP_UPDATE",
    ],
  ],
  Array [
    "66d83870-30df-4a3b-8801-59edff162034",
    Object {
      "user-id": "667802d8-2260-436c-958a-2ee0f71f73f0",
    },
    Array [
      "WEB_INTEGRATIONS_APP_UPDATE",
    ],
  ],
]
`;

exports[`delete-smooch-web-integration Everthing is successful sends back status 200 when the web integration is deleted successfully 1`] = `
Object {
  "body": Object {
    "deleted": true,
    "message": "The web integration with for tenant 66d83870-30df-4a3b-8801-59edff162034 and integrationId 667802d8-2260-436c-958a-2ee0f71f73f0 has been deleted successfully",
  },
  "status": 200,
}
`;

exports[`delete-smooch-web-integration sends back status 400 when there are invalid parameters 1`] = `
Object {
  "body": Object {
    "error": [ValidationError: "id" is required],
    "message": "Error: invalid params value(s). \\"id\\" is required",
  },
  "status": 400,
}
`;

exports[`delete-smooch-web-integration sends back status 403 when there are not enough permissions 1`] = `
Object {
  "body": Object {
    "expectedPermissions": Object {
      "tenant": Array [
        "WEB_INTEGRATIONS_APP_UPDATE",
      ],
    },
    "message": "Error not enough permissions",
  },
  "status": 403,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when the app does not exist for the tenant 1`] = `
Object {
  "body": Object {
    "message": "The app does not exist for this tenant",
  },
  "status": 400,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when there is a error deleting records in DynamoDB 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete a record in DynamoDB",
  },
  "status": 500,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when there is a error deleting web integration 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to delete an web integration",
  },
  "status": 500,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when there is a error fetching an app in DynamoDB 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to fetch an app in DynamoDB",
  },
  "status": 500,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when there is a error retrieving digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to retrieve digital channels credentials",
  },
  "status": 500,
}
`;

exports[`delete-smooch-web-integration sends back status 500 when there is a error validating digital channels credentials 1`] = `
Object {
  "body": Object {
    "message": "An Error has occurred trying to validate digital channels credentials",
  },
  "status": 500,
}
`;
